RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1634 @@
+          hcalFastCluster_singleSeed(
+              acc, pfClusParams, topoId, nRHTopo, pfRecHits, pfClusteringVars, clusterView, fracView);
+        } else if (nSeeds <= 100 && nRHTopo - nSeeds < threadsPerBlockForClustering) {
+          hcalFastCluster_multiSeedParallel(
+              acc, pfClusParams, topoId, nSeeds, nRHTopo, pfRecHits, pfClusteringVars, clusterView, fracView);
+        } else if (nSeeds <= 400 && nRHTopo - nSeeds <= 1500) {
+          hcalFastCluster_multiSeedIterative(
+              acc, pfClusParams, topoId, nSeeds, nRHTopo, pfRecHits, pfClusteringVars, clusterView, fracView);
+        } else {
+          if (once_per_block(acc))
+            printf("ERROR: Topo cluster %d has %d seeds and %d rechits. Will be processed in next kernel.\n",
+                   topoId,
+                   nSeeds,
+                   nRHTopo);
+        }
+      } else
+        return;
+    }
+  };
+
+  // Process very large, exotic topo clusters
+  class FastClusterExotic {
+  public:
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC void operator()(const TAcc& acc,
+                                  const reco::PFRecHitHostCollection::ConstView pfRecHits,
+                                  const reco::PFClusterParamsDeviceCollection::ConstView pfClusParams,
+                                  reco::PFClusteringVarsDeviceCollection::View pfClusteringVars,
+                                  reco::PFClusterDeviceCollection::View clusterView,
+                                  reco::PFRecHitFractionDeviceCollection::View fracView,
+                                  Position4* __restrict__ globalClusterPos,
+                                  Position4* __restrict__ globalPrevClusterPos,
+                                  float* __restrict__ globalClusterEnergy,
+                                  float* __restrict__ globalRhFracSum,
+                                  int* __restrict__ globalSeeds,
+                                  int* __restrict__ globalRechits) const {
+      const int nRH = pfRecHits.size();
+      int& topoId = alpaka::declareSharedVar<int, __COUNTER__>(acc);
+      int& blockId = alpaka::declareSharedVar<int, __COUNTER__>(acc);
+      int& nRHTopo = alpaka::declareSharedVar<int, __COUNTER__>(acc);
+      int& nSeeds = alpaka::declareSharedVar<int, __COUNTER__>(acc);
+      uint32_t& counter = alpaka::declareSharedVar<uint32_t, __COUNTER__>(acc);
+
+      if (once_per_block(acc)) {
+        counter = 0;
+        blockId = alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0u];
+      }
+      alpaka::syncBlockThreads(acc);  // all threads call sync
+
+      for (counter = 0;
+           counter < (((nRH + (blocksForExoticClusters - 1)) / blocksForExoticClusters) * blocksForExoticClusters);
+           counter += blocksForExoticClusters) {
+        if (once_per_block(acc)) {
+          topoId = blockId + counter;
+          if (topoId > nRH)
+            break;
```

Comment:
This part looks wrong ?

`topoId` is the same for all threads in the block, but only the first thread in the block will break out of the `for` loop.

Reply:
I put this in to mainly prevent any out of bounds access in the following two lines. Is there a better solution?

Reply:
The problem is that the `break` will exit from the `for (counter = 0; ...` loop, but only for one of the threads in the block, and this will break the `syncBlockThreads`.

I would make `topoId`, `nRHTopo` and `nSeeds` local variables instead of shared memory variables, and remove the ` if (once_per_block(acc))`.

Or, actually, go with the change suggested below.